File: python/test/mod_test_relative_axis_contributions.py
"""
Unit tests for relative axis contributions Python wrappers
"""
import numpy as np
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from tensoromics_functions import (
    relative_axes_changes_from_shift_vector,
    relative_axes_expression_from_expression_vector
)
from test_helpers import run_all_tests, assert_error

TOL = 1e-12


def assert_vec_sum(vec, expected_sum, tol=TOL, msg=""):
    actual_sum = np.sum(vec)
    assert np.isclose(actual_sum, expected_sum, atol=tol), f"{msg} Expected sum {expected_sum}, got {actual_sum}"


def assert_all_finite(vec, msg=""):
    assert np.all(np.isfinite(vec)), f"{msg} Non-finite values detected."


def assert_in_range(vec, min_val=0.0, max_val=1.0, msg=""):
    assert not np.any((vec < min_val) | (vec > max_val)), f"{msg} Values out of range [{min_val}, {max_val}]"


# Tests for relative_axes_changes_from_shift_vector
def test_shift_positive_vector():
    vec = np.array([1, 2, 3], dtype=np.float64)
    contrib = relative_axes_changes_from_shift_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="shift positive vector")
    assert_all_finite(contrib, "shift positive vector")
    assert_in_range(contrib, msg="shift positive vector")


def test_shift_negative_vector():
    vec = np.array([-1, -2, -3], dtype=np.float64)
    contrib = relative_axes_changes_from_shift_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="shift negative vector")
    assert_all_finite(contrib, "shift negative vector")
    assert_in_range(contrib, msg="shift negative vector")


def test_shift_mixed_vector():
    vec = np.array([2, -2, 4], dtype=np.float64)
    contrib = relative_axes_changes_from_shift_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="shift mixed vector")
    assert_all_finite(contrib, "shift mixed vector")
    assert_in_range(contrib, msg="shift mixed vector")


def test_shift_zero_vector():
    vec = np.array([0, 0, 0], dtype=np.float64)
    assert_error(lambda: relative_axes_changes_from_shift_vector(vec), "Expected exception for zero vector")


def test_shift_one_nonzero_axis():
    vec = np.array([0, 5, 0], dtype=np.float64)
    contrib = relative_axes_changes_from_shift_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="shift one nonzero axis")
    assert_all_finite(contrib, "shift one nonzero axis")
    assert_in_range(contrib, msg="shift one nonzero axis")


def test_shift_all_equal():
    vec = np.array([2, 2, 2], dtype=np.float64)
    contrib = relative_axes_changes_from_shift_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="shift all equal")
    assert_all_finite(contrib, "shift all equal")
    assert_in_range(contrib, msg="shift all equal")


def test_shift_large_vector():
    vec = np.ones(100, dtype=np.float64)
    contrib = relative_axes_changes_from_shift_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="shift large vector")
    assert_all_finite(contrib, "shift large vector")
    assert_in_range(contrib, msg="shift large vector")


# Tests for relative_axes_expression_from_expression_vector
def test_expr_positive_vector():
    vec = np.array([1, 2, 3], dtype=np.float64)
    contrib = relative_axes_expression_from_expression_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="expr positive vector")
    assert_all_finite(contrib, "expr positive vector")
    assert_in_range(contrib, msg="expr positive vector")


def test_expr_negative_vector():
    vec = np.array([-1, -2, -3], dtype=np.float64)
    contrib = relative_axes_expression_from_expression_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="expr negative vector")
    assert_all_finite(contrib, "expr negative vector")
    assert_in_range(contrib, msg="expr negative vector")


def test_expr_mixed_vector():
    vec = np.array([2, -2, 4], dtype=np.float64)
    contrib = relative_axes_expression_from_expression_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="expr mixed vector")
    assert_all_finite(contrib, "expr mixed vector")
    assert_in_range(contrib, msg="expr mixed vector")


def test_expr_zero_vector():
    vec = np.array([0, 0, 0], dtype=np.float64)
    assert_error(lambda: relative_axes_expression_from_expression_vector(vec), "Expected exception for zero vector")


def test_expr_one_nonzero_axis():
    vec = np.array([0, 5, 0], dtype=np.float64)
    contrib = relative_axes_expression_from_expression_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="expr one nonzero axis")
    assert_all_finite(contrib, "expr one nonzero axis")
    assert_in_range(contrib, msg="expr one nonzero axis")


def test_expr_all_equal():
    vec = np.array([2, 2, 2], dtype=np.float64)
    contrib = relative_axes_expression_from_expression_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="expr all equal")
    assert_all_finite(contrib, "expr all equal")
    assert_in_range(contrib, msg="expr all equal")


def test_expr_large_vector():
    vec = np.ones(100, dtype=np.float64)
    contrib = relative_axes_expression_from_expression_vector(vec)
    assert_vec_sum(contrib, 1.0, msg="expr large vector")
    assert_all_finite(contrib, "expr large vector")
    assert_in_range(contrib, msg="expr large vector")


if __name__ == "__main__":
    run_all_tests(globals().values())

File: r/test/mod_test_relative_axis_contributions.R
## ==================== TESTS FOR RELATIVE AXIS CONTRIBUTIONS ====================
# Load the compiled Fortran library
source("r/load_tensor_omics.R")
source("r/test_helpers.R")

#' Assert that all values in a vector are in [0, 1]
assert_in_range <- function(vec, min_val = 0, max_val = 1, message = "") {
  assert_false(any(vec < min_val | vec > max_val), sprintf("Test failed: %s. Values out of range [%g, %g]", message, min_val, max_val))
}

#' Assert that the sum of a vector is approximately equal to a value
assert_sum_equal <- function(vec, expected_sum, tolerance = 1e-12, message = "") {
  actual_sum <- sum(vec)
  assert_equal_numeric(actual_sum, expected_sum, sprintf("Test failed: %s. Expected sum %f, got %f", message, expected_sum, actual_sum))
}


#' Test: relative_axes_changes_from_shift_vector
test_shift_positive_vector <- function() {
  vec <- c(1, 2, 3)
  contrib <- relative_axes_changes_from_shift_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "shift positive vector: sum")
  assert_in_range(contrib, message = "shift positive vector: range")
}

test_shift_negative_vector <- function() {
  vec <- c(-1, -2, -3)
  contrib <- relative_axes_changes_from_shift_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "shift negative vector: sum")
  assert_in_range(contrib, message = "shift negative vector: range")
}

test_shift_mixed_vector <- function() {
  vec <- c(2, -2, 4)
  contrib <- relative_axes_changes_from_shift_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "shift mixed vector: sum")
  assert_in_range(contrib, message = "shift mixed vector: range")
}

test_shift_zero_vector <- function() {
  # as above: no direction to apportion, so this is an error, not zeros
  vec <- c(0, 0, 0)
  assert_error(relative_axes_changes_from_shift_vector(vec),
               "Expected an error for a zero shift vector", ERR_DIVISION_BY_ZERO)
}

test_shift_one_nonzero_axis <- function() {
  vec <- c(0, 5, 0)
  contrib <- relative_axes_changes_from_shift_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "shift one nonzero axis: sum")
  assert_in_range(contrib, message = "shift one nonzero axis: range")
}

test_shift_all_equal <- function() {
  vec <- c(2, 2, 2)
  contrib <- relative_axes_changes_from_shift_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "shift all equal: sum")
  assert_in_range(contrib, message = "shift all equal: range")
}

test_shift_large_vector <- function() {
  vec <- rep(1, 100)
  contrib <- relative_axes_changes_from_shift_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "shift large vector: sum")
  assert_in_range(contrib, message = "shift large vector: range")
}

#' Test: relative_axes_expression_from_expression_vector
test_expr_positive_vector <- function() {
  vec <- c(1, 2, 3)
  contrib <- relative_axes_expression_from_expression_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "expr positive vector: sum")
  assert_in_range(contrib, message = "expr positive vector: range")
}

test_expr_negative_vector <- function() {
  vec <- c(-1, -2, -3)
  contrib <- relative_axes_expression_from_expression_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "expr negative vector: sum")
  assert_in_range(contrib, message = "expr negative vector: range")
}

test_expr_mixed_vector <- function() {
  vec <- c(2, -2, 4)
  contrib <- relative_axes_expression_from_expression_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "expr mixed vector: sum")
  assert_in_range(contrib, message = "expr mixed vector: range")
}

test_expr_zero_vector <- function() {
  # a zero vector has no direction to apportion, so the Fortran reports a division by
  # zero rather than returning zeros. The Python suite asserts the same.
  vec <- c(0, 0, 0)
  assert_error(relative_axes_expression_from_expression_vector(vec),
               "Expected an error for a zero expression vector", ERR_DIVISION_BY_ZERO)
}

test_expr_one_nonzero_axis <- function() {
  vec <- c(0, 5, 0)
  contrib <- relative_axes_expression_from_expression_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "expr one nonzero axis: sum")
  assert_in_range(contrib, message = "expr one nonzero axis: range")
}

test_expr_all_equal <- function() {
  vec <- c(2, 2, 2)
  contrib <- relative_axes_expression_from_expression_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "expr all equal: sum")
  assert_in_range(contrib, message = "expr all equal: range")
}

test_expr_large_vector <- function() {
  vec <- rep(1, 100)
  contrib <- relative_axes_expression_from_expression_vector(vec)
  assert_sum_equal(contrib, 1.0, message = "expr large vector: sum")
  assert_in_range(contrib, message = "expr large vector: range")
}

run_all_tests()
